src/b3d/chisight/gen3d/image_kernel.py

```diff
@@ -310,3 +310,153 @@ def get_rgbd_vertex_kernel(self) -> PixelRGBDDistribution:
+        unique_pixel_coordinates, unique_indices = jnp.unique(
+            projected_coords,
+            axis=0,
+            return_index=True,
+            size=projected_coords.shape[0],
+            fill_value=INVALID_IDX,
+        )
+
+        observed_rgbd_per_point = observed_rgbd.at[
+            unique_pixel_coordinates[:, 0], unique_pixel_coordinates[:, 1]
+        ].get(mode="drop", fill_value=-1.0)
+
+        # For each collided pixel, get the color and depth of the vertex that hit that pixel.
+        latent_rgbd_per_point = jnp.concatenate(
+            [
+                state["colors"][sort_order][unique_indices],
+                transformed_points_sorted[unique_indices, 2, None],
+            ],
+            axis=-1,
+        )
+
+        is_valid = (
+            (unique_pixel_coordinates >= 0).all(axis=-1)
+            * (observed_rgbd_per_point[..., 3] >= 0.0),
+        )
+        return (
+            observed_rgbd_per_point,
+            latent_rgbd_per_point,
+            is_valid,
+            unique_pixel_coordinates,
+        )
+
+    def logpdf(
+        self, observed_rgbd: FloatArray, state: Mapping, hyperparams: Mapping
+    ) -> FloatArray:
+        intrinsics = hyperparams["intrinsics"]
+        (observed_rgbd_per_point, latent_rgbd_per_point, is_valid, _) = (
+            self.calculate_latent_and_observed_correspondences(
+                observed_rgbd, state, hyperparams
+            )
+        )
+
+        # Score the collided pixels
+        scores = jax.vmap(
+            hyperparams["image_kernel"].get_rgbd_vertex_kernel().logpdf,
+            in_axes=(0, 0, None, None, 0, 0, None),
+        )(
+            observed_rgbd_per_point,
+            latent_rgbd_per_point,
+            state["color_scale"],
+            state["depth_scale"],
+            state["visibility_prob"],
+            state["depth_nonreturn_prob"],
+            hyperparams["intrinsics"],
+        )
+        total_score_for_explained_pixels = jnp.where(is_valid, scores, 0.0).sum()
+
+        # Score the pixels that don't have any vertices hitting them.
+        number_of_pixels_with_no_hypothesis = (
+            hyperparams["intrinsics"]["image_height"].unwrap()
+            * hyperparams["intrinsics"]["image_width"].unwrap()
+        ) - is_valid.sum()
+
+        number_of_total_non_return_pixels = (observed_rgbd[..., 3] == 0.0).sum()
+        number_of_non_return_pixels_with_latent_hypothesis = (
+            observed_rgbd_per_point[..., 3] == 0.0
+        ).sum()
+        number_of_non_return_pixels_without_latent_hypothesis = (
+            number_of_total_non_return_pixels
+            - number_of_non_return_pixels_with_latent_hypothesis
+        )
+
+        # The pixels that have no vertices produce an observation uniformly at random.
+        color_score = jnp.log(1 / 1.0**3)
+
+        depth_score_return = jnp.log(1 - state["depth_nonreturn_prob"]) + jnp.log(
```

Comment:
@nishadgothoskar this looks great!  Is `state["depth_nonreturn_prob"]` the vector of depth nonreturn probs associated with each point?  If so, I think we want this to be a different value -- something like `hyperparams["depth_nonreturn_prob_for_unexplained_pixel"]`.

Reply:
e.g. https://github.com/probcomp/b3d/blob/main/src/b3d/chisight/gen3d/pixel_kernels/pixel_depth_kernels.py#L22